lib/modem_slm/modem_slm.c

```diff
@@ -205,109 +256,218 @@ static void parse_at_response(const char *data, size_t datalen)
+	if (err) {
+		LOG_ERR("UART TX error: %d", err);
+		ring_buf_get_finish(&tx_buf, 0);
+		return err;
+	}
+
+	return 0;
 }
 
-static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+/* Write the data to tx_buffer and trigger sending. */
+static int tx_write(const uint8_t *data, size_t len, bool flush)
 {
+	size_t ret;
+	size_t sent = 0;
 	int err;
-	static uint16_t pos;
-	static bool enable_rx_retry;
 
-	ARG_UNUSED(dev);
-	ARG_UNUSED(user_data);
+	LOG_HEXDUMP_DBG(data, len, "TX");
+
+	while (sent < len) {
+		ret = ring_buf_put(&tx_buf, data + sent, len - sent);
+		if (ret) {
+			sent += ret;
+		} else {
+			/* Buffer full, block and start TX. */
+			k_sem_take(&tx_done_sem, K_FOREVER);
+			err = tx_start();
+			if (err) {
+				LOG_ERR("TX buf overflow, %d dropped. Unable to send: %d",
+					len - sent,
+					err);
+				k_sem_give(&tx_done_sem);
+				return err;
+			}
+		}
+	}
+
+	if (flush && k_sem_take(&tx_done_sem, K_NO_WAIT) == 0) {
+		err = tx_start();
+		if (err == 1) {
+			k_sem_give(&tx_done_sem);
+			return 0;
+		} else if (err) {
+			LOG_ERR("tx_start failed: %d", err);
+			k_sem_give(&tx_done_sem);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static void uart_callback(const struct device*, struct uart_event *evt, void*)
+{
+	struct rx_buf_t *buf;
+	struct rx_event_t rx_event;
+	int err;
 
 	switch (evt->type) {
 	case UART_TX_DONE:
-		k_free(uart_tx_buf);
-		k_sem_give(&tx_done);
-		LOG_DBG("TX_DONE");
-		break;
+		err = ring_buf_get_finish(&tx_buf, evt->data.tx.len);
+		if (err) {
+			LOG_ERR("UART_TX_DONE failure: %d", err);
+		}
+		if (ring_buf_is_empty(&tx_buf) == false) {
+			tx_start();
+		} else {
+			k_sem_give(&tx_done_sem);
+		}
+	break;
 	case UART_TX_ABORTED:
-		k_free(uart_tx_buf);
-		k_sem_give(&tx_done);
-		LOG_INF("TX_ABORTED");
-		break;
+		ring_buf_get_finish(&tx_buf, evt->data.tx.len);
+		LOG_WRN("TX_ABORTED, dropped: %d bytes", ring_buf_size_get(&tx_buf));
+		ring_buf_reset(&tx_buf);
+		k_sem_give(&tx_done_sem);
+	break;
 	case UART_RX_RDY:
-		LOG_DBG("RX_RDY %d", evt->data.rx.len);
-		uart_rx_handler(&(evt->data.rx.buf[pos]), evt->data.rx.len);
-		pos += evt->data.rx.len;
-		break;
+		rx_buf_ref(evt->data.rx.buf);
+		rx_event.buf = &evt->data.rx.buf[evt->data.rx.offset];
+		rx_event.len = evt->data.rx.len;
+		err = k_msgq_put(&rx_event_queue, &rx_event, K_NO_WAIT);
+		if (err) {
+			LOG_ERR("RX_RDY failure: %d, dropped: %d", err, evt->data.rx.len);
+			rx_buf_unref(evt->data.rx.buf);
+			break;
+		}
+		k_work_submit((struct k_work *)&rx_process_work);
+	break;
 	case UART_RX_BUF_REQUEST:
-		LOG_DBG("RX_BUF_REQ");
-		pos = 0;
-		err = uart_rx_buf_rsp(uart_dev, next_buf, sizeof(uart_rx_buf[0]));
+		buf = rx_buf_alloc();
+		if (!buf) {
+			break;
+		}
+		err = uart_rx_buf_rsp(uart_dev, buf->buf, sizeof(buf->buf));
 		if (err) {
-			LOG_WRN("UART RX buf rsp: %d", err);
+			LOG_WRN("Disabling UART RX: %d", err);
+			rx_buf_unref(buf);
```

Comment:
`rx_buf_alloc()` sets `ref_count` to `1`. If an error happens, `rx_buf_unref()` decreases the value to `0` and then checks if the value is `1`. It isn't, so the block is not freed. Unless `uart_rx_buf_rsp()` indirectly causes `rx_buf_ref()` to be called?